::::{tip} Laboratory Task 2
:::{div}
:style: "font-family:Times New Roman; text-align:justify; font-size:15px"

<b>Instruction:</b> Perform a single forward pass and compute for the error.
:::

$$x = \begin{bmatrix} 1 \\ 0 \\ 1 \end{bmatrix}$$

$$y = \begin{bmatrix} 1 \end{bmatrix}$$

$$f=max(0, Z_n)$$

$$\text{hidden unit weights} =
\begin{bmatrix}
w_{11} = 0.2 && w_{12} = -0.3 \\
w_{13} = 0.4 && w_{14} = 0.1 \\
w_{15} = -0.5 && w_{16} = 0.2
\end{bmatrix}$$

$$\text{output unit weights} =
\begin{bmatrix}
w_{21} = -0.3 \\
w_{22} = -0.2
\end{bmatrix}$$

$$\theta =
\begin{bmatrix}
\theta_{1} = -0.4 \\
\theta_{2} = 0.2 \\
\theta_{3} = 0.1
\end{bmatrix}$$
::::

:::{div}
:style: "font-family:Times New Roman; text-align:justify; font-size:15px"

This is the same setup used in the sigmoid example earlier in the course notebook, except the activation
function here is ReLU instead of sigmoid. Since ReLU just clips negative values to zero, most of the
arithmetic below is the same weighted-sum-plus-bias formula from before, just passed through <tt>max(0, z)</tt>
instead of the sigmoid curve.
:::

<h4 style="font-family: Times New Roman"><strong>Step 1: Set Up the Given Values</strong></h4>

In [1]:
import numpy as np

# input vector and target output
x = np.array([1, 0, 1])
y = 1

# hidden layer weights (input -> hidden)
w11, w12 = 0.2, -0.3
w13, w14 = 0.4, 0.1
w15, w16 = -0.5, 0.2

# output layer weights (hidden -> output)
w21, w22 = -0.3, -0.2

# biases
theta1, theta2, theta3 = -0.4, 0.2, 0.1

print("x =", x)
print("y =", y)

x = [1 0 1]
y = 1


<h4 style="font-family: Times New Roman"><strong>Step 2: Define the ReLU Activation</strong></h4>

:::{div}
:style: "font-family:Times New Roman; text-align:justify; font-size:15px"

The task specifies $f = max(0, Z_n)$, which is just ReLU. If the weighted sum is negative, the unit outputs 0;
otherwise it passes the value through unchanged.
:::

In [2]:
def relu(z):
    return max(0, z)

<h4 style="font-family: Times New Roman"><strong>Step 3: Compute the Weighted Sums Going into the Hidden Layer</strong></h4>

:::{div}
:style: "font-family:Times New Roman; text-align:justify; font-size:15px"

Following the same pattern shown earlier in the course notebook:
:::

$$Z_1 = (w_{11} \cdot x_1) + (w_{13} \cdot x_2) + (w_{15} \cdot x_3) + \theta_1$$
$$Z_2 = (w_{12} \cdot x_1) + (w_{14} \cdot x_2) + (w_{16} \cdot x_3) + \theta_2$$

In [3]:
Z1 = (w11 * x[0]) + (w13 * x[1]) + (w15 * x[2]) + theta1
Z2 = (w12 * x[0]) + (w14 * x[1]) + (w16 * x[2]) + theta2

print("Z1 =", Z1)
print("Z2 =", Z2)

Z1 = -0.7
Z2 = 0.10000000000000003


<h4 style="font-family: Times New Roman"><strong>Step 4: Pass Z1 and Z2 through ReLU to Get the Hidden Layer Outputs</strong></h4>

:::{div}
:style: "font-family:Times New Roman; text-align:justify; font-size:15px"

$$H_1 = f(Z_1) \qquad H_2 = f(Z_2)$$
Since $Z_1 = -0.7$ is negative, ReLU kills it and $H_1$ becomes 0. $Z_2 = 0.1$ is positive so it just stays as
it is.
:::

In [4]:
H1 = relu(Z1)
H2 = relu(Z2)

print("H1 =", H1)
print("H2 =", H2)

H1 = 0
H2 = 0.10000000000000003


<h4 style="font-family: Times New Roman"><strong>Step 5: Compute the Weighted Sum for the Output Layer</strong></h4>

:::{div}
:style: "font-family:Times New Roman; text-align:justify; font-size:15px"

$$Z_3 = (w_{21} \cdot H_1) + (w_{22} \cdot H_2) + \theta_3$$
$$\hat{y} = f(Z_3)$$
:::

In [5]:
Z3 = (w21 * H1) + (w22 * H2) + theta3
y_hat = relu(Z3)

print("Z3 =", Z3)
print("y_hat =", y_hat)

Z3 = 0.08
y_hat = 0.08


<h4 style="font-family: Times New Roman"><strong>Step 6: Compute the Error</strong></h4>

:::{div}
:style: "font-family:Times New Roman; text-align:justify; font-size:15px"

There are two common ways to express the error here, so both are shown below.
:::

:::{div}
:style: "font-family:Times New Roman; text-align:justify; font-size:15px"

1. The simple/signed error, useful for the backpropagation step covered in the next laboratory task:
$$\delta = \hat{y} - y$$
:::

:::{div}
:style: "font-family:Times New Roman; text-align:justify; font-size:15px"

2. The squared error, which is the more standard "loss" formulation:
$$E = \frac{1}{2}(y - \hat{y})^2$$
:::

In [6]:
delta = y_hat - y
squared_error = 0.5 * (y - y_hat) ** 2

print("delta (y_hat - y) =", delta)
print("squared error      =", squared_error)

delta (y_hat - y) = -0.92
squared error      = 0.4232


<h4 style="font-family: Times New Roman"><strong>Summary</strong></h4>

<center>
<table style="display: inline-block; font-family: Times New Roman">
<tr><th>Quantity</th><th>Value</th></tr>
<tr><td>$Z_1$</td><td>-0.7</td></tr>
<tr><td>$Z_2$</td><td>0.1</td></tr>
<tr><td>$H_1$</td><td>0</td></tr>
<tr><td>$H_2$</td><td>0.1</td></tr>
<tr><td>$Z_3$</td><td>0.08</td></tr>
<tr><td>$\hat{y}$</td><td>0.08</td></tr>
<tr><td>$\delta = \hat{y} - y$</td><td>-0.92</td></tr>
<tr><td>$E = \frac{1}{2}(y-\hat{y})^2$</td><td>0.4232</td></tr>
</table>
</center>

:::{div}
:style: "font-family:Times New Roman; text-align:justify; font-size:15px"

The prediction ($\hat{y} = 0.08$) is way off from the target ($y = 1$), which makes sense since the weights
were just randomly picked and haven't been trained at all. $H_1$ ended up being exactly 0 because ReLU zeroed
out the negative $Z_1$, meaning that particular hidden unit contributed nothing to the final output on this
pass. This large error is exactly what backpropagation would use afterward to update the weights and push
$\hat{y}$ closer to 1.
:::

In [7]:
print(f"Forward pass result: y_hat = {y_hat}")
print(f"Target: y = {y}")
print(f"Error (delta): {delta}")
print(f"Squared error: {squared_error}")

Forward pass result: y_hat = 0.08
Target: y = 1
Error (delta): -0.92
Squared error: 0.4232
